### PR TITLE
Option to exclude groups from workflow pickup [BW-1061]

### PR DESCRIPTION
**Pull request by aednichols**

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala
Comment (aednichols) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala:176]: Since it's a bit hard to see amid all the noise, this is where the next ticket should start passing in actual groups

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala
Comment (aednichols) [inline database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala:82]: I need to verify that removing this query from `Compiled` status doesn't introduce unacceptable performance degradation.

It is a known issue in Slick [0] that `inSet` and `Compiled` do not get along.

[0] https://github.com/slick/slick/issues/718

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala
Comment (aednichols) [inline database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala:97]: This is the one line that really actually changed

**Comment by aednichols** [comment]
Resolving merge conflicts...

**Review by salonishah11 — APPROVED** [review]
LGTM 👍

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala
Comment (cjllanwarne) [reply database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala:82]: What form will this verification take?

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/InMemoryWorkflowStore.scala
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/InMemoryWorkflowStore.scala:35]: Does the "in memory workflow store" respect this new parameter?

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala
Comment (cjllanwarne) [reply database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala:82]: What form will this verification take?

File: engine/src/test/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStoreSpec.scala
Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStoreSpec.scala:217]: For completeness (not that I doubt this), it would be nice to check that "fetch 10 excluding X" does the right thing (ie fetches 10 workflows that don't match X, rather than fetching 10 workflows and THEN filtering out the Xs)

**Review by cjllanwarne — COMMENTED** [review]
LGTM pending one nice-to-have test and more details on that plan for testing the performance.

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala
Comment (aednichols) [reply database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala:82]: My plan is to add a case to the suite that calls `workflowStore.fetchStartableWorkflows` as quickly as possible in a loop for a fixed number of iterations and measure its execution time. I think we have microbenchmarks like this elsewhere in the codebase. Didn't get to it today due to various UL fires.

The target is somewhat lenient, the query is executed at most once per second per instance: https://github.com/broadinstitute/firecloud-develop/blob/dev/base-configs/cromwell/cromwell.conf.ctmpl#L133-L139

Comment (aednichols) [reply database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala:82]: Alternatively I would love to find a solution that keeps `Compiled`, I bashed my head against it unsuccessfully for a good while last week.

Comment (aednichols) [reply database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala:82]: Wrote a speed measurement test case [0] and ran it  this branch and `develop`.

Time is seconds to execute the query 500 times.

Non-compiled:
5.9, 5.6, 5.5
Avg. 5.7 seconds

Compiled:
4.0, 3.9, 4.1
Avg. 4.0 seconds

So there is a noticeable difference, but it's not one that I'm worried about when considering the current configured rate of 1 query per second.

[0]

```
it should "be acceptably fast" taggedAs DbmsTest in {
  (for {
    submissionResponsesExcluded <- workflowStore.add(sourceFilesCollection)
    _ = (for (i <- 1 to 500) {
      Await.result(workflowStore.fetchStartableWorkflows(10, "A08", 1.second), 1.second)
    })
  } yield ()).futureValue
}
```

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/InMemoryWorkflowStore.scala
Comment (aednichols) [reply engine/src/main/scala/cromwell/engine/workflow/workflowstore/InMemoryWorkflowStore.scala:35]: It currently accepts it and drops it on the ground. The in-memory store is used for testing and for the single workflow runner, and I don't think it's applicable in either place. And the [underlying `var workflowStore`](https://github.com/broadinstitute/cromwell/blob/develop/engine/src/main/scala/cromwell/engine/workflow/workflowstore/InMemoryWorkflowStore.scala#L16) does not track group in an accessible manner. 

I don't feel amazing about how the interface pattern forces the in-memory version to include it in the param list as if it actually did something, but I can't think of anything better at present.

Comment (cjllanwarne) [reply engine/src/main/scala/cromwell/engine/workflow/workflowstore/InMemoryWorkflowStore.scala:35]: Sounds good to me. If you felt very civic minded you could add an explicit "throw new NotImplementedException" if it's not empty, but I'll leave that up to you

**Review by cjllanwarne — APPROVED** [review]

**Comment by aednichols** [comment]
Before:
```
select `WORKFLOW_EXECUTION_UUID`, `WORKFLOW_DEFINITION`, `WORKFLOW_URL`, `WORKFLOW_ROOT`, `WORKFLOW_TYPE`, `WORKFLOW_TYPE_VERSION`, `WORKFLOW_INPUTS`, `WORKFLOW_OPTIONS`, `WORKFLOW_STATE`, `SUBMISSION_TIME`, `IMPORTS_ZIP`, `CUSTOM_LABELS`, `CROMWELL_ID`, `HEARTBEAT_TIMESTAMP`, `HOG_GROUP`, `WORKFLOW_STORE_ENTRY_ID` from `WORKFLOW_STORE_ENTRY` where ((`HEARTBEAT_TIMESTAMP` is null) or (`HEARTBEAT_TIMESTAMP` < '2022-02-01 21:13:48.287')) and (not (`WORKFLOW_STATE` = 'On Hold')) order by `SUBMISSION_TIME` limit 1 for update
```
After, with no groups excluded: `and (not false)`
```
select `WORKFLOW_EXECUTION_UUID`, `WORKFLOW_DEFINITION`, `WORKFLOW_URL`, `WORKFLOW_ROOT`, `WORKFLOW_TYPE`, `WORKFLOW_TYPE_VERSION`, `WORKFLOW_INPUTS`, `WORKFLOW_OPTIONS`, `WORKFLOW_STATE`, `SUBMISSION_TIME`, `IMPORTS_ZIP`, `CUSTOM_LABELS`, `CROMWELL_ID`, `HEARTBEAT_TIMESTAMP`, `HOG_GROUP`, `WORKFLOW_STORE_ENTRY_ID` from `WORKFLOW_STORE_ENTRY` where (((`HEARTBEAT_TIMESTAMP` is null) or (`HEARTBEAT_TIMESTAMP` < '2022-03-01 20:08:12.447')) and (not (`WORKFLOW_STATE` = 'On Hold'))) and (not false) order by `SUBMISSION_TIME` limit 1 for update
```

**Pull request merged**
